### ISSUE TITLE
Weirdly-specific type narowing for bool-tagged unions

### ISSUE THREAD

**Issue by Ontonator** [microsoft/pyright#4875]
**Describe the bug**
When using a tagged union of tuples with a `bool` tag, it can only be narrowed using `==`, not `is` or directly using the value of the `bool`.

**To Reproduce**
Create a variable `tagged_union: Tuple[Literal[True], int] | Tuple[Literal[False], str]` and narrow with:
- `if tagged_union[0]:` (narrows `tagged_union[0]` but not `tagged_union`)
- `if tagged_union[0] is True:` (narrows `tagged_union[0]` but not `tagged_union`)
- `if tagged_union[0] == True:` (narrows `tagged_union[0]` and `tagged_union`)

**Expected behavior**
The type of `tagged_union` should be narrowed in all three cases.

**Screenshots or Code**
```python
from typing import Tuple, Literal


def test(tagged_union: Tuple[Literal[True], int] | Tuple[Literal[False], str]) -> None:
    # tagged_union not narrowed
    if tagged_union[0]:
        reveal_type(tagged_union)  # Tuple[Literal[True], int] | Tuple[Literal[False], str]
        reveal_type(tagged_union[0])  # Literal[True]
    else:
        reveal_type(tagged_union)  # Tuple[Literal[True], int] | Tuple[Literal[False], str]
        reveal_type(tagged_union[0])  # Literal[False]

    # tagged_union not narrowed
    if tagged_union[0] is True:
        reveal_type(tagged_union)  # Tuple[Literal[True], int] | Tuple[Literal[False], str]
        reveal_type(tagged_union[0])  # Literal[True]
    else:
        reveal_type(tagged_union)  # Tuple[Literal[True], int] | Tuple[Literal[False], str]
        reveal_type(tagged_union[0])  # Literal[False]

    # tagged_union narrowed
    if tagged_union[0] == True:
        reveal_type(tagged_union)  # Tuple[Literal[True], int]
        reveal_type(tagged_union[0])  # Literal[True]
    else:
        reveal_type(tagged_union)  # Tuple[Literal[False], str]
        reveal_type(tagged_union[0])  # Literal[False]
```

**VS Code extension or command-line**
Running version 1.1.301 as language server with Neovim and as command line tool.

**Comment by erictraut** [comment]
Yes, type narrowing logic is very specific to the form in which it appears. The logic for each form is necessarily different because each operator has subtle differences in how it works. 

Please refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards) for details.

Interestingly, mypy does apply narrowing in the two cases where pyright does not, and it doesn't apply narrowing in the case where it does.

I'll take a deeper look at the code to see if there's a straightforward way to extend narrowing to the cases above. The `is True` form should be straightforward. The simple truthy/falsy check might be more problematic; that check is very performance sensitive, so if there's a measurable performance impact to type analysis generally, I don't think it's worth adding support for this very specific type narrowing case.

**Comment by erictraut** [comment]
I've added support for the form `x[I] is B` and `x[I] is not B` (where `B` is a bool literal). The reason this was not previously supported is that other literal types (like `int` or `str` literals) cannot be used in this manner because the `is` operator is not guaranteed to work for literal values other than `bool` and enums, so the `==` operator is much more commonly used when testing for tagged tuples.

As I suspected, the performance impact of supporting narrowing for the more general truthy/falsy case is quite high. Adding this support would slow down analysis in the general case, and I can't justify that for such a niche type narrowing use case. You'll need to use the `==` or `is` operators if you want pyright to apply type narrowing in this case.
